libcaf_net/caf/net/web_socket/framing.cpp

```diff
@@ -246,11 +159,141 @@ bool framing::end_text_message() {
+  return hdr_bytes;
+}
+
+ptrdiff_t framing::consume_payload(byte_span buffer, byte_span delta) {
+  // Calculate at what point of the received buffer the delta payload begins.
+  auto offset = static_cast<ptrdiff_t>(buffer.size() - delta.size());
+  // Unmask the arrived data.
+  if (hdr_.mask_key != 0)
+    detail::rfc6455::mask_data(hdr_.mask_key, buffer, offset);
+
+  // Handle control frames first, since these may not me fragmented,
+  // and can arrive between regular message fragments.
+  if (detail::rfc6455::is_control_frame(hdr_.opcode)) {
+    return handle(hdr_.opcode, buffer, hdr_.payload_len);
+  }
+  // Handle the fragmentation logic of text and binary messages.
+  if (hdr_.opcode == detail::rfc6455::text_frame
+      || opcode_ == detail::rfc6455::text_frame) {
+    // For text messages we validate the UTF-8 encoding on the go. Only text
+    // messages can arrive with incomplete payload.
+    if (hdr_.opcode == detail::rfc6455::text_frame && hdr_.fin) {
+      if (!payload_valid(buffer, validation_offset_)) {
+        abort_and_shutdown(sec::malformed_message, "Invalid UTF-8 sequence");
+        return -1;
+      }
+    } else {
+      payload_buf_.insert(payload_buf_.end(), buffer.begin() + offset,
+                          buffer.end());
+      if (!payload_valid(payload_buf_, validation_offset_)) {
+        abort_and_shutdown(sec::malformed_message, "Invalid UTF-8 sequence");
+        return -1;
+      }
+    }
+    // Wait for more data if necessary.
+    if (buffer.size() < hdr_.payload_len)
+      return 0;
+  } else if ((hdr_.opcode == detail::rfc6455::binary_frame && !hdr_.fin)
+             || opcode_ == detail::rfc6455::binary_frame) {
+    payload_buf_.insert(payload_buf_.end(), buffer.begin(), buffer.end());
+  }
+  // Handle the completed frame.
+  if (hdr_.fin) {
+    if (opcode_ == detail::rfc6455::invalid_frame) {
+      if (hdr_.opcode == detail::rfc6455::text_frame
+          && validation_offset_ != buffer.size()) {
+        abort_and_shutdown(sec::malformed_message, "Invalid UTF-8 sequence");
+        return -1;
+      }
+      // Call upper layer.
+      validation_offset_ = 0;
+      return handle(hdr_.opcode, buffer, hdr_.payload_len);
+    }
+    // End of fragmented input.
+    if (opcode_ == detail::rfc6455::text_frame
+        && validation_offset_ != payload_buf_.size()) {
+      abort_and_shutdown(sec::malformed_message, "Invalid UTF-8 sequence");
+      return -1;
+    }
+    auto result = handle(opcode_, payload_buf_, hdr_.payload_len);
+    opcode_ = detail::rfc6455::invalid_frame;
+    payload_buf_.clear();
+    validation_offset_ = 0;
+    return result;
+  }
+  if (opcode_ == detail::rfc6455::invalid_frame)
+    opcode_ = hdr_.opcode;
+  // Clean up the state since we finished processing this frame
+  down_->configure_read(receive_policy::up_to(2048));
+  hdr_.opcode = detail::rfc6455::invalid_frame;
+  return static_cast<ptrdiff_t>(hdr_.payload_len);
```

Comment:
This code is not DRY: the same lines are copied at the end of the function. Could be factored out into a lambda, e.g.:

```
auto receive_next_header = [this] {
  down_->configure_read(receive_policy::up_to(2048));
  hdr_.opcode = detail::rfc6455::invalid_frame;
  return static_cast<ptrdiff_t>(hdr_.payload_len);
};
```

And then these lines would just become `return receive_next_header();`.